llpc/context/llpcCompiler.cpp

```diff
@@ -571,6 +572,253 @@ template <typename T> static FormattedBytes formatBytesLittleEndian(ArrayRef<T>
+// Build the whole graphics pipeline. If missing elfPackage of a certain stage, we will build it first, and
+// link them to the full pipeline last.
+//
+// @param pipelineInfo : Info to build this shader module
+// @param [out] pipelineOut : Output of building this shader module
+// @param elfPackage : Early compiled elfPackage; it is an array which size is UnlinkedStageCount.
+// @returns : Result::Success if successful. Other return codes indicate failure.
+Result Compiler::buildGraphicsPipelineWithElf(const GraphicsPipelineBuildInfo *pipelineInfo,
+                                              GraphicsPipelineBuildOut *pipelineOut, const BinaryData *elfPackage) {
+  if (!pipelineInfo->pfnOutputAlloc)
+    return Result::ErrorInvalidPointer;
+
+  // clang-format off
+  SmallVector<const PipelineShaderInfo *, ShaderStageGfxCount> shaderInfo = {
+    &pipelineInfo->vs,
+    &pipelineInfo->tcs,
+    &pipelineInfo->tes,
+    &pipelineInfo->gs,
+    &pipelineInfo->fs,
+  };
+  // clang-format on
+
+  MetroHash::Hash cacheHash = {};
+  MetroHash::Hash pipelineHash = {};
+  cacheHash = PipelineDumper::generateHashForGraphicsPipeline(pipelineInfo, true, false);
+  pipelineHash = PipelineDumper::generateHashForGraphicsPipeline(pipelineInfo, false, false);
+
+  Optional<CacheAccessor> cacheAccessor;
+  if (cl::CacheFullPipelines) {
+    cacheAccessor.emplace(pipelineInfo, cacheHash, getInternalCaches());
+  }
+
+  Result result = Result::Success;
+  BinaryData elfBin = {};
+  ElfPackage elf[UnlinkedStageCount] = {};
+  ElfPackage pipelineElf;
+  if (!cacheAccessor && cacheAccessor->isInCache()) {
+    LLPC_OUTS("Cache hit for graphics pipeline.\n");
+    elfBin = cacheAccessor->getElfFromCache();
+    pipelineOut->pipelineCacheAccess =
+        cacheAccessor->hitInternalCache() ? CacheAccessInfo::InternalCacheHit : CacheAccessInfo::CacheHit;
+  } else {
+    GraphicsContext graphicsContext(m_gfxIp, pipelineInfo, &pipelineHash, &cacheHash);
+    Context *context = acquireContext();
+    context->attachPipelineContext(&graphicsContext);
+
+    // Release context.
+    auto onExit = make_scope_exit([&] { releaseContext(context); });
+
+    for (unsigned idx = 0; idx < UnlinkedStageCount; idx++) {
+      if (elfPackage[idx].pCode) {
+        auto data = reinterpret_cast<const char *>(elfPackage[idx].pCode);
+        elf[idx].assign(data, data + elfPackage[idx].codeSize);
+      } else {
+        graphicsContext.setUnlinked(true);
+        result = buildUnlinkedShaderInternal(context, shaderInfo, static_cast<UnlinkedShaderStage>(idx), elf[idx],
+                                             pipelineOut->stageCacheAccesses);
+        if (result != Result::Success)
+          return result;
+      }
+    }
+
+    // Link
+    graphicsContext.setUnlinked(false);
+    bool hasError = false;
+    context->setDiagnosticHandler(std::make_unique<LlpcDiagnosticHandler>(&hasError));
+    hasError |= !linkRelocatableShaderElf(elf, &pipelineElf, context);
```

Comment:
I think I mentioned this before.  You should call `Compiler::canUseRelocatableGraphicsShaderElf` at some point early in this function.  There are some cases that will generate incorrect code if you try to link the unlinked elf.  That function lets you know.